// vue-api-query/build/Parser.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _qs = _interopRequireDefault(require("qs"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var Parser = /*#__PURE__*/function () {
  function Parser(builder) {
    (0, _classCallCheck2["default"])(this, Parser);
    this.builder = builder;
    this.uri = '';
  } // final query string


  (0, _createClass2["default"])(Parser, [{
    key: "query",
    value: function query() {
      this.reset();
      this.includes();
      this.appends();
      this.fields();
      this.filters();
      this.sorts();
      this.page();
      this.limit();
      this.payload();
      return this.uri;
    }
  }, {
    key: "reset",
    value: function reset() {
      this.uri = '';
    }
    /**
     * Helpers
     */

  }, {
    key: "hasIncludes",
    value: function hasIncludes() {
      return this.builder.includes.length > 0;
    }
  }, {
    key: "hasAppends",
    value: function hasAppends() {
      return this.builder.appends.length > 0;
    }
  }, {
    key: "hasFields",
    value: function hasFields() {
      return Object.keys(this.builder.fields).length > 0;
    }
  }, {
    key: "hasFilters",
    value: function hasFilters() {
      return Object.keys(this.builder.filters).length > 0;
    }
  }, {
    key: "hasSorts",
    value: function hasSorts() {
      return this.builder.sorts.length > 0;
    }
  }, {
    key: "hasPage",
    value: function hasPage() {
      return this.builder.pageValue !== null;
    }
  }, {
    key: "hasLimit",
    value: function hasLimit() {
      return this.builder.limitValue !== null;
    }
  }, {
    key: "hasPayload",
    value: function hasPayload() {
      return this.builder.payload !== null;
    }
  }, {
    key: "prepend",
    value: function prepend() {
      return this.uri === '' ? '?' : '&';
    }
  }, {
    key: "parameterNames",
    value: function parameterNames() {
      return this.builder.model.parameterNames();
    }
    /**
     * Parsers
     */

  }, {
    key: "includes",
    value: function includes() {
      if (!this.hasIncludes()) {
        return;
      }

      this.uri += this.prepend() + this.parameterNames().include + '=' + this.builder.includes;
    }
  }, {
    key: "appends",
    value: function appends() {
      if (!this.hasAppends()) {
        return;
      }

      this.uri += this.prepend() + this.parameterNames().append + '=' + this.builder.appends;
    }
  }, {
    key: "fields",
    value: function fields() {
      if (!this.hasFields()) {
        return;
      }

      var fields = (0, _defineProperty2["default"])({}, this.parameterNames().fields, this.builder.fields);
      this.uri += this.prepend() + _qs["default"].stringify(fields, _objectSpread({
        encode: false
      }, this.builder.model.stringifyOptions()));
    }
  }, {
    key: "filters",
    value: function filters() {
      if (!this.hasFilters()) {
        return;
      }

      var filters = (0, _defineProperty2["default"])({}, this.parameterNames().filter, this.builder.filters);
      this.uri += this.prepend() + _qs["default"].stringify(filters, _objectSpread({
        encode: false
      }, this.builder.model.stringifyOptions()));
    }
  }, {
    key: "sorts",
    value: function sorts() {
      if (!this.hasSorts()) {
        return;
      }

      this.uri += this.prepend() + this.parameterNames().sort + '=' + this.builder.sorts;
    }
  }, {
    key: "page",
    value: function page() {
      if (!this.hasPage()) {
        return;
      }

      this.uri += this.prepend() + this.parameterNames().page + '=' + this.builder.pageValue;
    }
  }, {
    key: "limit",
    value: function limit() {
      if (!this.hasLimit()) {
        return;
      }

      this.uri += this.prepend() + this.parameterNames().limit + '=' + this.builder.limitValue;
    }
  }, {
    key: "payload",
    value: function payload() {
      if (!this.hasPayload()) {
        return;
      }

      this.uri += this.prepend() + _qs["default"].stringify(this.builder.payload, _objectSpread({
        encode: false
      }, this.builder.model.stringifyOptions()));
    }
  }]);
  return Parser;
}();

exports["default"] = Parser;
